### ISSUE TITLE
mask handler does not behave as expected

### ISSUE THREAD

**Issue by dsheldon** [pyro-ppl/numpyro#568]
I am trying to use the mask handler to mask invalid observed values (nans). Per [this discussion](https://forum.pyro.ai/t/behavior-of-mask-handler-with-invalid-observation-possible-bug/1719), I replace invalid observations with values that are valid for the particular distribution to avoid nans in the gradient. But now it seems that the default values actually affect the posterior distribution.

Please see this example, where x ~ N(0, 1) and y_i ~ N(x, 0.1) for i = 1 to n:

~~~python
import jax
import jax.numpy as np
from jax.random import PRNGKey

import numpyro
import numpyro.distributions as dist
from numpyro.infer import MCMC, NUTS

def model(n, y=None):
    
    x = numpyro.sample("x", dist.Normal(0, 1))
    
    mask = True 
    if y is not None:
        # Mask nan values, then substitute a default value to
        # avoid nans in gradients
        mask = ~np.isnan(y)
        y = np.where(mask, y, -100) 
        
    with numpyro.handlers.mask(mask):
        y = numpyro.sample("y", dist.Normal(x, 0.1), sample_shape=(n,), obs=y)
        
    return x, y


n = 100
with numpyro.handlers.seed(rng_seed=1):
    x, _ = model(n)
    print("x is", x)
    
y = np.repeat(np.nan, n)

kernel = NUTS(model)
mcmc = MCMC(kernel, num_warmup=1000, num_samples=1000, num_chains=1)
mcmc.run(PRNGKey(1), n, y=y)

mcmc.print_summary()
~~~

Here I provide an observed vector `y` that is all `nan`, so I expect the posterior over x to be equal to the prior. However, the posterior I get is highly concentrated about the default value (-100) used to replace nans:

~~~text
                mean       std    median      5.0%     95.0%     n_eff     r_hat
         x    -99.99      0.01    -99.99   -100.01    -99.97    410.46      1.01
~~~

I'm filing this as an issue because I believe that masked observations are not supposed to affect the log-density or its gradients, but let me know if I've misinterpreted something.

**Comment by fehiepsi** [comment]
Your argument sounds very reasonable to me. Let me look at this in more details.

**Comment by fehiepsi** [comment]
@dsheldon It seems that you miss the keyword `mask_array`:
```
with numpyro.handlers.mask(mask_array=mask):
```

**Comment by dsheldon** [comment]
Oh, wow. I guess I misunderstood how to call this.

**Comment by dsheldon** [comment]
Wait, does inference work for you when you make this change? Now I get:
~~~
         x       nan       nan       nan       nan       nan       nan       nan
~~~

**Comment by dsheldon** [comment]
Update, with this change (using `mask_arrray=mask`), it seems to work as long as there it at least *one* non-masked value:
~~~python
n = 10
with numpyro.handlers.seed(rng_seed=1):
    x, y_true = model(n)
    print("x is", x)

for k in [0, 1, 5]:
    y = np.where(np.arange(n) < k, y_true, np.nan)
    print("y is ", y)
    
    kernel = NUTS(model)
    mcmc = MCMC(kernel, num_warmup=1000, num_samples=1000, num_chains=1)
    mcmc.run(PRNGKey(2), n, y=y)

    mcmc.print_summary()
~~~

For k=0 (all nans) inference diverges and gives all nans as in my last post. For k=1 and k=5 it seems to behave as expected.

**Comment by fehiepsi** [comment]
Thanks! It eventually leads to another bug. ;)

**Comment by dsheldon** [comment]
Thanks!! I picked up your fix and this makes a big difference in the model I'm working on.

**Comment by fehiepsi** [comment]
> this makes a big difference in the model I'm working on

glad to hear! :)
